common/dictionary-db/dictionary-db.ts

```diff
@@ -0,0 +1,1416 @@
+            .where('[profile+noteId]')
+            .anyOf(noteIds.map((noteId) => [profile, noteId]))
+            .toArray()
+            .then((ankiCards) => {
+                if (!ankiCards.length) return new Map();
+                const noteIdRecordMap = new Map<number, DictionaryAnkiCardRecord[]>();
+                for (const ankiCard of ankiCards) {
+                    const val = noteIdRecordMap.get(ankiCard.noteId);
+                    if (val) val.push(ankiCard);
+                    else noteIdRecordMap.set(ankiCard.noteId, [ankiCard]);
+                }
+                return noteIdRecordMap;
+            });
+    }
+
+    private async _saveRecordBulk(records: DictionaryTokenRecord[]): Promise<DictionaryTokenKey[]> {
+        if (!records.length) return [];
+        return this.db.tokens.bulkPut(records, { allKeys: true });
+    }
+
+    /**
+     * There are four scenarios where tokens/cards need to be deleted:
+     * 1. The card was removed from Anki (handled by _syncTrackStatesWithAnki())
+     * 2. The card field was removed/renamed (handled by _syncTrackStatesWithAnki())
+     * 3. The card field value no longer produce the same tokens (handled by _saveTokensForDB())
+     * 4. Based on track settings such as no Anki fields (handled by tracksToClear)
+     */
+    private async _deleteCardBulk(
+        profile: string,
+        orphanedTrackCardIds: Map<number, number[]>,
+        modifiedTokens: Set<string>
+    ): Promise<void> {
+        for (const [track, cardIds] of orphanedTrackCardIds.entries()) {
+            if (!cardIds.length) orphanedTrackCardIds.delete(track);
+        }
+        if (!orphanedTrackCardIds.size) return;
+
+        return this.db.transaction('rw', this.db.tokens, this.db.ankiCards, async () => {
+            await mapAsync(Array.from(orphanedTrackCardIds.entries()), ([track, orphanedCardIds]) => {
+                const cardIdsSet = new Set(orphanedCardIds);
+                return Promise.all([
+                    this.db.tokens
+                        .where('cardIds')
+                        .anyOf(orphanedCardIds)
+                        .distinct()
+                        .filter((r) => r.track === track && r.profile === profile)
+                        .modify((record, ref) => {
+                            const remainingCardIds = record.cardIds.filter((id) => !cardIdsSet.has(id));
+                            if (!remainingCardIds.length) {
+                                modifiedTokens.add(record.token);
+                                for (const lemma of record.lemmas) modifiedTokens.add(lemma);
+                                delete (ref as any).value;
+                            } else if (remainingCardIds.length !== record.cardIds.length) {
+                                modifiedTokens.add(record.token);
+                                for (const lemma of record.lemmas) modifiedTokens.add(lemma);
+                                record.cardIds = remainingCardIds;
```

Comment:
Minor nitpick - seems like you'd get more code re-use and readability if you only branch for the final modification to the record.